Source/DafnyDriver/DafnyDoc.cs

```diff
@@ -0,0 +1,1142 @@
+      var decls = rootModule.ModuleDef.TopLevelDecls.Select(d => !(d is LiteralModuleDecl));
+      CollectDecls(rootModule, modDecls);
+      WriteTOC(modDecls);
+      foreach (var m in modDecls) {
+        WriteModule(m, dafnyFiles);
+      }
+      WriteIndex();
+      return DafnyDriver.ExitValue.SUCCESS;
+    } catch (Exception e) {
+      // This is a fail-safe backstop so that dafny itself does not crash
+      Reporter.Error(MessageSource.Documentation, DafnyProgram.DefaultModule, $"Unexpected exception while generating documentation: {e.Message}\n{e.StackTrace}");
+      return DafnyDriver.ExitValue.DAFNY_ERROR;
+    }
+  }
+
+  /** Recursively computes a list of all the LiteralModuleDecls declared in the program */
+  public void CollectDecls(LiteralModuleDecl mod, List<LiteralModuleDecl> modDecls) {
+    modDecls.Add(mod);
+    foreach (var d in mod.ModuleDef.TopLevelDecls) {
+      if (d is LiteralModuleDecl litmod) {
+        CollectDecls(litmod, modDecls);
+      }
+    }
+  }
+
+  /** Writes a doc page for the given module */
+  public void WriteModule(LiteralModuleDecl module, IList<DafnyFile> dafnyFiles) {
+    var moduleDef = module.ModuleDef;
+    var fullName = moduleDef.FullDafnyName;
+    var fullNLName = fullName;
+    if (moduleDef.IsDefaultModule) {
+      nameIndex.Add(RootNLName + " " + nameIndex.Count + " " + RootName, "module " + Link(RootName, RootNLName));
+      fullName = RootName;
+      fullNLName = RootNLName;
+    } else {
+      AddToIndexF(module.Name, fullName, "module");
+    }
+    var defaultClass = moduleDef.TopLevelDecls.First(d => d is ClassDecl cd && cd.IsDefaultClass) as ClassDecl;
+
+    string filename = Outputdir + "/" + fullName + ".html";
+    using StreamWriter file = new(filename);
+    file.Write(HtmlStart($"Module {fullName} in program {DafnyProgram.Name}"));
+    file.Write(style);
+    var abs = moduleDef.IsAbstract ? "abstract " : ""; // The only modifier for modules
+    file.WriteLine(Heading1($"{abs}module {QualifiedNameWithLinks(fullNLName, false)}{space4}{Smaller(contentslink + indexlink)}"));
+    file.Write(BodyStart());
+
+    var docstring = Docstring(module);
+    if (!String.IsNullOrEmpty(docstring)) {
+      file.Write(ShortAndMoreForDecl(module));
+      file.Write(br);
+      file.WriteLine(br);
+    }
+    if (moduleDef.RefinementQId != null) {
+      file.WriteLine("refines " + QualifiedNameWithLinks(moduleDef.RefinementQId.Decl.FullDafnyName) + br);
+    }
+    var attributes = module.Attributes?.ToString();
+    if (!String.IsNullOrEmpty(attributes)) {
+      file.WriteLine("Attributes: " + attributes + br);
+    }
+
+    if (moduleDef.IsDefaultModule) {
+      if (dafnyFiles.Count != 1) {
+        file.WriteLine("From multiple files<br>\n");
+      } else {
+        file.WriteLine(FileInfo(dafnyFiles[0].CanonicalPath));
+      }
+    } else {
+      file.Write(FileInfo(module.Tok));
+    }
+
+    StringBuilder summaries = new StringBuilder(1000);
```

Comment:
How do you decide what's in summaries and what's in details? Is the initialization with `1000` useful?

Reply:
Summaries are intended to be one-line per declaration for a quick view of the contents. Details give the full signature and docstrings and specs.

The 1000 is purely arbitrary. Just seems like the default of 16 is too small and causes gratuitous resizes.